cody-repo-util.el

```diff
@@ -0,0 +1,182 @@
+      (setq remote-url (cody--url-do-replacements remote-url))
+      (setq remote-branch-name (cody--get-remote-branch-name project file))
+      (when (or (not remote-branch-name)
+                (not (string-match-p remote-branch-name remote-url)))
+        (setq remote-branch-name cody-default-branch-name)))
+    (list :vcs-type vcs-type
+          :remote-url remote-url
+          :remote-branch-name (or remote-branch-name cody--default-branch-name)
+          :relative-path relative-path)))
+
+(defun cody--find-repository-name (project current-file)
+  "Find the repository name for PROJECT and CURRENT-FILE."
+  (when-let ((file-from-repo (or current-file
+                                 (cody--get-root-file-from-first-git-repo project))))
+    (condition-case nil
+        (cody--get-remote-repo-url project file-from-repo)
+      (error
+       (cody--get-simple-repository-name project file-from-repo)))))
+
+(defun cody--get-simple-repository-name (project file)
+  "Get the simple repository name for PROJECT and FILE."
+  (when-let ((repository (vc-call-backend (vc-backend file) 'root file)))
+    (file-name-nondirectory (directory-file-name repository))))
+
+(defun cody--url-do-replacements (remote-url)
+  "Perform URL replacements for REMOTE-URL"
+  (let* ((replacements
+          (split-string cody--remote-url-replacements "\\s*,\\s*"))
+         (remote-url-with-replacements remote-url))
+    (when (zerop (% (length replacements) 2))
+      (cl-loop for i from 0 to (1- (/ (length replacements) 2))
+               do (setq remote-url-with-replacements
+                        (replace-regexp-in-string
+                         (nth (* 2 i) replacements)
+                         (nth (1+ (* 2 i)) replacements)
+                         remote-url-with-replacements))))
+    remote-url-with-replacements))
+
+(defun cody--get-remote-repo-url (project file)
+  "Get the remote repository URL for PROJECT and FILE.
+Returned format: github.com:sourcegraph/sourcegraph.git"
+  (let ((repository (vc-call-backend (vc-backend file) 'root file))
+        (vcs-type (cody--get-vcs-type project file)))
+    (cond
+     ((and (eq vcs-type 'git) repository)
+      (cody--convert-git-clone-url-to-codebase-name (vc-git-repository-url repository)))
+     ((eq vcs-type 'perforce)
+      (cody--get-perforce-remote-repo-url project file))
+     (t (error "Unsupported VCS or repository not found for file %s" file)))))
+
+(defun cody--get-repo-root-path (project file)
+  "Get the repository root directory for PROJECT and FILE."
+  (let ((vcs-root (vc-call-backend (vc-backend file) 'root file)))
+    (when vcs-root
+      (file-name-directory vcs-root))))
+
+(defun cody--get-remote-branch-name (project file)
+  "Get the remote branch name for PROJECT and FILE."
+  (let ((repository (vc-call-backend (vc-backend file) 'root file)))
+    (when (and repository (eq (cody--get-vcs-type project file) 'git))
+      (vc-git-working-revision repository))))
```

Comment:
Is this really a remote branch name? Doesn't this just do `git rev-parse HEAD`, which is a commit hash? And it seems like `vc-git-working-revision` doesn't care about its filename argument? (...at least with the one I have with emacs-28.2.)

Reply:
fixed